RedfishPkg/RedfishHttpDxe/RedfishHttpDxe.c

```diff
@@ -0,0 +1,1336 @@
+      gBS->Stall (Private->RetrySetting.RetryWait);
+    }
+  } while (TRUE);
+
+  //
+  // Redfish resource is updated. Automatically expire the cached response
+  // so application can directly get resource from Redfish service again.
+  //
+  DEBUG ((REDFISH_HTTP_CACHE_DEBUG, "%a: Resource is updated, expire URI: %s\n", __func__, Uri));
+  RedfishExpireResponse (This, Uri);
+
+  if (EFI_ERROR (Status)) {
+    DEBUG_CODE (
+      DumpRedfishResponse (NULL, DEBUG_ERROR, Response);
+      );
+    //
+    // Report status code for Redfish failure
+    //
+    ReportHttpError (HttpMethodDelete, Uri, Response->StatusCode);
+    DEBUG ((DEBUG_ERROR, "%a: delete %s failed (%d/%d): %r\n", __func__, Uri, RetryCount, Private->RetrySetting.MaximumRetryDelete, Status));
+    goto ON_RELEASE;
+  }
+
+ON_RELEASE:
+
+  return Status;
+}
+
+EDKII_REDFISH_HTTP_PROTOCOL  mEdkIIRedfishHttpProtocol = {
+  EDKII_REDFISH_HTTP_PROTOCOL_REVISION,
+  RedfishCreateRedfishService,
+  RedfishFreeRedfishService,
+  RedfishJsonInRedfishPayload,
+  RedfishGetResource,
+  RedfishPatchResource,
+  RedfishPutResource,
+  RedfishPostResource,
+  RedfishDeleteResource,
+  RedfishFreeRequest,
+  RedfishFreeResponse,
+  RedfishExpireResponse
+};
+
+/**
+  Unloads an image.
+
+  @param[in]  ImageHandle         Handle that identifies the image to be unloaded.
+
+  @retval EFI_SUCCESS             The image has been unloaded.
+  @retval EFI_INVALID_PARAMETER   ImageHandle is not a valid image handle.
+
+**/
+EFI_STATUS
+EFIAPI
+RedfishHttpDriverUnload (
+  IN EFI_HANDLE  ImageHandle
+  )
+{
+  if (mRedfishHttpCachePrivate == NULL) {
+    return EFI_SUCCESS;
+  }
+
+  if (!IsListEmpty (&mRedfishHttpCachePrivate->CacheList.Head)) {
+    ReleaseCacheList (&mRedfishHttpCachePrivate->CacheList);
+  }
+
+  gBS->UninstallMultipleProtocolInterfaces (
+         ImageHandle,
+         &gEdkIIRedfishHttpProtocolGuid,
+         &mRedfishHttpCachePrivate->Protocol,
+         NULL
+         );
+
+  FreePool (mRedfishHttpCachePrivate);
+  mRedfishHttpCachePrivate = NULL;
+
+  return EFI_SUCCESS;
+}
+
+/**
+  This is a EDKII_REDFISH_CREDENTIAL_PROTOCOL notification event handler.
+
+  @param[in] Event    Event whose notification function is being invoked.
+  @param[in] Context  Pointer to the notification function's context.
+
+**/
+VOID
+EFIAPI
+CredentialProtocolInstalled (
+  IN  EFI_EVENT  Event,
+  IN  VOID       *Context
+  )
+{
+  EFI_STATUS                  Status;
+  REDFISH_HTTP_CACHE_PRIVATE  *Private;
+
+  Private = (REDFISH_HTTP_CACHE_PRIVATE *)Context;
+  if (Private->Signature != REDFISH_HTTP_DRIVER_SIGNATURE) {
+    return;
+  }
+
+  //
+  // Locate HII database protocol.
+  //
+  Status = gBS->LocateProtocol (
+                  &gEdkIIRedfishCredentialProtocolGuid,
+                  NULL,
+                  (VOID **)&Private->CredentialProtocol
+                  );
+  if (EFI_ERROR (Status)) {
+    return;
+  }
+
+  gBS->CloseEvent (Event);

```

Comment:
We can move this to the beginning of funciton.

Reply:
No, I think we only close event when we can successfully locate protocol.

Reply:
My thought is when we get the notification of Redfish HTTP protocol installation. We should be able to locate the protocol. If not, then there must be something wrong. Suppose we won't get another notification of protocol installation as there is only one instance of Redfish HTTP protocol, right?